### PR TITLE
Arnej/add network ports 1

### PR DESCRIPTION
**Pull request by arnej27959**
@mpolden please review (especially config-provisioning and node-repository changes)
@bratseth please review
@hmusum FYI

### CODE REVIEW

File: config-provisioning/src/main/java/com/yahoo/config/provision/Provisioner.java [hunk 1]
```diff
@@ -25,6 +25,13 @@
      */
     List<HostSpec> prepare(ApplicationId applicationId, ClusterSpec cluster, Capacity capacity, int groups, ProvisionLogger logger);
 
+    /**
+     * fetch or create a port allocator for the given host
+     * @param host the host spec
+     * @return the port allocator
+     **/
+    default NetworkPorts getNetworkPorts(HostSpec host) { return new NetworkPorts(); }
```
Comment (bratseth) [inline config-provisioning/src/main/java/com/yahoo/config/provision/Provisioner.java]: I don't think this should be a separate method on this level, but part of the payload inside HostSpec such that existing ports are returned as part of the hosts in prepare, and the wanted complete list of ports is passed back to the node repo in activate.

**Review by bratseth — COMMENTED** [review]
Node, and everything it owns must be immutable, but it is probably not practical to have this immutable on the config model side? I think you need to split NetworkPorts into an immutable and mutable implementation and use one in Node and the other in the config model. 

Only allocated nodes have ports so I think this should be under Allocation rather than directly on Node.

File: config-model/src/main/java/com/yahoo/vespa/model/ConfigProxy.java [hunk 2]
```diff
@@ -47,6 +47,13 @@ public ConfigProxy(Host host) {
      */
     public int getPortCount() { return 1; }
 
+    @Override
+    public String[] getPortSuffixes() {
+        String[] suffixes = new String[1];
```
Comment (mpolden) [inline config-model/src/main/java/com/yahoo/vespa/model/ConfigProxy.java]: Can be simplified to a single statement, `return new String[]{"rpc"}`. Applies to other methods that do this as well.

File: config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Optional;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Contains information about a port (port number and a collection of tags).
+ *
+ */
+public class NetworkPorts {
+
+    public static Optional<NetworkPorts> empty() { return Optional.empty(); }
```
Comment (mpolden) [inline config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java]: Unnecessary?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java [hunk 4]
```diff
@@ -101,6 +105,7 @@ public Node(String id, Set<String> ipAddresses, Set<String> ipAddressPool, Strin
         this.type = type;
         this.reports = reports;
         this.modelId = modelId;
+        this.networkPorts = networkPorts;
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java]: Wrap in `Objects.requireNonNull` like other fields.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java [hunk 5]
```diff
@@ -125,6 +133,17 @@ private void toSlime(Node node, Cursor object) {
         node.status().firmwareVerifiedAt().ifPresent(instant -> object.setLong(firmwareCheckKey, instant.toEpochMilli()));
         node.reports().toSlime(object, reportsKey);
         node.modelId().ifPresent(modelId -> object.setString(modelIdKey, modelId));
+        node.networkPorts().ifPresent(ports -> toSlime(ports, object.setArray(networkPortsKey)));
+    }
+
+    private void toSlime(NetworkPorts networkPorts, Cursor array) {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java]: `SerializationTest` should be updated to test serialization of these new fields.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java [hunk 6]
```diff
@@ -265,6 +276,17 @@ private Allocation requireAllocation(String message) {
         return allocation.get();
     }
 
+    /** Return the current port allocator for this node, if any */
+    public Optional<NetworkPorts> networkPorts() { return networkPorts; }
+
+    /** Return a network port allocator for this node */
+    public NetworkPorts getNetworkPorts() {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java]: Is this used? In any case, I think it's better for the caller to deal with constructing NetworkPorts if there is none.

**Comment by arnej27959** [comment]
i have added some more commits here, still missing unit tests (at least).
@bratseth and @mpolden, can you please take a quick look to see if it's going in the right direction?  Will make a new PR (rebased from master branch) tomorrow.

File: config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java [hunk 7]
```diff
@@ -47,56 +44,14 @@ public String toString() {
...
-    public Map<Integer, Allocation> byPortMap() {
-        return ImmutableMap.copyOf(byPorts);
-    }
-    public Map<String, Allocation> byKeyMap() {
-        return ImmutableMap.copyOf(byKeys);
-    }
-    public Collection<Allocation> allocations() {
-        return ImmutableMap.copyOf(byPorts).values();
+    public NetworkPorts(Collection<Allocation> allocations) {
+        this.allocations = new ArrayList<>(allocations.size());
```
Comment (mpolden) [inline config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java]: Can be simplified to `new ArrayList<>(allocations)`, or simply `List.copyOf` which also omits the need for `unmodifiableList` wrapping below.

**Review by mpolden — COMMENTED** [review]
Looks good so far. I think it would also be nice to expose these ports in the nodes API, for easy inspection. See `NodesResponse`.

File: config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java (hunk 7 above)
Comment (bratseth) [reply config-provisioning/src/main/java/com/yahoo/config/provision/NetworkPorts.java]: Or just ImmutableList also in the field declaration, which makes it explicit that the list won't change.

**Comment by bratseth** [comment]
This looks good now

**Pull request closed without merging**